### ISSUE TITLE
Order by columns not work properly

### ISSUE THREAD

**Issue by garenk02** [n1crack/datatables#33]
When I try to do sort per column it not return the correct sort (asc/desc)

**Comment by seancpdev** [comment]
I have a similar issue actually. I am getting a larger data set imported to narrow down the issue. I know on my implementation, I have several joins in the MySQL statement. So not sure if that plays a factor or not. But I noticed at least one or two rows which would not remain in order. Like if I sort dates, there would be a date that would remain out of order.

**Comment by n1crack** [comment]
Can I see your sql query string?
And which db you are using?

**Comment by seancpdev** [comment]
MySQL

SELECT reg.id, reg.payment_method as type, reg.source, reg.firstname, reg.lastname, reg.email, 
                evt.id as evtid, evt.title, loc.id as locid, loc.city, loc.state, evt.start, reg.status, trans
                .promocode_discount, trans.promocode,
                trans
                .groupdiscount, 
                trans.earlybird, (trans.amount/trans.tickets_purchased) as amount FROM event_registration as reg 
                LEFT JOIN events as evt ON reg.event = evt.id LEFT JOIN event_locations as loc ON evt.location = loc.id 
                LEFT JOIN transactions as trans ON reg.parent = trans.uid 
                LEFT JOIN event_categories as cat ON evt.category = cat.id

**Comment by n1crack** [comment]
That seem to be something related joins. 

Can you try changing join type ? And try again later ?

**Comment by seancpdev** [comment]
I updated the JOINS and it didn't help. I did, however, discover the issue. Several of the selects are used primarily for logic display and are later hidden using your framework code. 

I realized if I keep the 'hidden' selects at the end of the query i.e. SELECT id, type, firstname, lastname, email, hidden.field, hidden.field the sorting started working correctly. Guessing the columns are assigned in order on the query, so any hidden columns need to be set last. 

So not sure best way to document for anyone else encountering a similar issue, but that appears to be what resolved it for me. I haven't merged my primary database yet, but will report back if I see my resolution didn't work on the full dataset.

**Comment by n1crack** [comment]
hmm.. I see. As you say there is a bug in hide() method or near it. 
Can you test https://github.com/n1crack/datatables/tree/2.0-dev  branch? It might have already fixed.

**Comment by seancpdev** [comment]
I don't think the bug is in the hide method unless that connects to the ordering. Unfortunately not too much time to review all the commits at the moment. I did see this: $this->columns->getByIndex($order['column'])->isOrderable(); in the commit. It would appear that possibly you're retrieving the index by the order of the 'columns' MySQL retrieves. Since the column ordering is done left to right in the SELECT statement, my guess is the columns are established in the same way -> left to right-> index 0, 1, 2, 3...

The columns themselves display as normal, as when I 'hide' them datatables doesn't try to pull them and I can use the data as part of edit or add functions as designed. The issue seems to be, it doesn't seem to reverse index them, at least on my version. Meaning if you select them in an arbitrary order with hidden columns between the 'displayed' ones, when ordering it uses the 'hidden column' as the basis for it's ordering.

So as stated, the 'visible' columns I added to the far left in the order they display and then any hidden columns I amended to the end of the SELECT portion of the statement.
